Sources/_StringProcessing/Engine/Processor.swift

```diff
@@ -688,18 +638,29 @@ extension Processor {
   }
 }
 
+// MARK: String matchers
+//
+// TODO: Refactor into separate file, formalize patterns
+
 extension String {
 
   func match(
     _ char: Character,
     at pos: Index,
-    limitedBy end: String.Index
+    limitedBy end: String.Index,
+    isCaseInsensitive: Bool
   ) -> Index? {
     // TODO: This can be greatly sped up with string internals
     // TODO: This is also very much quick-check-able
     assert(end <= endIndex)
 
-    guard pos < end, self[pos] == char else { return nil }
+    guard pos < end else { return nil }
+
+    if isCaseInsensitive {
+      guard self[pos].lowercased() == char.lowercased() else { return nil }
```

Comment:
Seems like we should:
(1) store `char` as lowercased so that we don't have to do it every time
(2) think about whether we can reduce the number of times we have to lowercase `self[pos]`

Reply:
Oh that's right, `lowercased()` and `properties.lowercaseMapping` both return strings, so this is a bit more complex.

Reply:
Agreed, and then we would assert it is that way (and use the argument label to mention this assumption). For 2, do you mean quick checks, or what?

Reply:
Also, if the known character to match is case-invariant, we can compile it to a case-sensitive comparison.